terraform/modules/gateway-google-cloud-compute/main.tf

```diff
@@ -46,6 +46,20 @@ locals {
     {
       name  = "FIREZONE_API_URL"
       value = var.api_url
+    },
+    # TODO: this is really confusing to set here, having explicit ID
+    # doesn't play well with horizontal replicas
+    {
```

Comment:
@jamilbk I was feeling confused here for some time thinking what to do with this ID, I bet clients trying to replicate this horizontally will have a hard time too

Reply:
Not that it has to change in this PR, but couldn't you just use the ID as the "base" and then append an integer at the point of use by using a terraform count index?

Reply:
@bmanifold there is no count for the instance template - we create just one, it doesn't matter how many actual instances are deployed from it. The same will happen with the AWS Instance Template and K8S Deployment yaml file.

So this decision to require `FIREZONE_ID` will force anyone doing it "the right way" writing custom environment-specific scripts trying to come up with a good way to uniquely identify an instance. 

I've used `hostname` as a hack but this is not the best idea, and it is pretty complicated if you think about it as a DevOps. Overall we will have to write 3-10x the amount of documentation and deployment examples just to keep it, here is a short intro:

## Kubernetes

For k8s the best way will be something like this:
```
- name: FIREZONE_ID
   valueFrom:
     fieldRef:
       fieldPath: spec.nodeName
```
pod names are unique so this might suffice, but on any deployment (or node upgrades) new pods will be created so new gateways will be created anyway, and this doesn't solve the issue of having a large amount of gateways. (So no charging per gateway is possible.)

The solution might be to enforce people using Stateful Sets for k8s (that have pods with static indexed names) but it only works if you want to have a bastion-like deployment. 

We can also use Daemon Sets and read actual node name but this will deploy a gateway to every node in a cluster, which can or can not be desirable either.

To sum up: gateways will be created with each pod (eg. sidecar deployment or just using regular deployment to create a replica set for our gateways) or we must enforce very strict deployment practices.

## Google Cloud

We will need to use a curl request and make sure that the instance service account token can access the metadata endpoint:
```
FIREZONE_ID=$(curl "http://metadata.google.internal/computeMetadata/v1/instance/id" -H "Metadata-Flavor: Google" -s)
# this will be set to 5895196805685327019
```
or
```
curl "http://metadata.google.internal/computeMetadata/v1/instance/hostname" -H "Metadata-Flavor: Google" -s
# this will be set to api-mftf.us-east1-d.c.firezone-staging.internal
```
but both will create new gateways on every deployment action. I don't know what is a good way to reliably solve it here other than not using Instance Templates / Groups and deploy barebone VMs that are harder to manage.

## AWS

Same as Google Cloud:
```
TOKEN=`curl -X PUT "http://169.254.169.254/latest/api/token" -q -H "X-aws-ec2-metadata-token-ttl-seconds: 21600"` \
&& curl -H "X-aws-ec2-metadata-token: $TOKEN" http://169.254.169.254/latest/meta-data/mac
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100    56  100    56    0     0  29787      0 --:--:-- --:--:-- --:--:-- 56000
12:ce:38:0b:f8:a1
```

Reply:
@AndrewDryga, yeah, totally understand that if you wanted to use AWS Auto-scaling groups (or GCP equivalent) my suggestion of the Terraform count index wouldn't work.

That said, have we had any customers asking/wanting to use the deployment methods you've described?  This is probably a question and topic better discussed in real-time rather rather than async.

Reply:
Yes, let's discuss it on a standup again.

I believe that GCP Instance Group is the recommended way to manage instances, no matter how many you want to deploy. Because it's the only way to get zero-downtime deployments, health checks, and automatic recovery. I think on AWS situation will be similar. It's also an easier way to manage a set of gateways (when there is more than one) - if you have more why would you use individual instances at all?..

Reply:
> Because it's the only way to get zero-downtime deployments, health checks, cross-az availability and automatic recovery

I think I'd disagree that it's the only way to accomplish these things.  But, as we've said, let's discuss in real-time

Reply:
cc @jamilbk every I try need to deploy firezone somewhere not as "one-off container" trying to prepare for the demo this is a huge pain in the ass - trying to put a bash script into a container environment variable.

Reply:
Hm, maybe I'm missing something, but why can't the customer just explicitly define their gateways' `FIREZONE_ID`'s like they explicitly define other terraform resources?

They'd be required to use a unique ID for each gateway, but I don't think that causes a problem for the foreseeable future based on expected use cases of having 3x gateways per site. It'd probably map to how they'll be thinking of gateways initially as infrastructure elements.

In other words, I'm not seeing why the ID needs to be dynamically generated, hence the motivation behind just using `TOKEN` to identify the gateway.

Reply:
Because it is impossible to do for pretty much anything (see examples above) except if you provision VMs one by one. Most idiomatic way to deploy anything these days are Instance Groups, K8S deployments, etc and they all are enforcing same environment variables for all the replicas.

Managing instances one by one is pain the ass if you need any advanced features (eg. zero downtime upgrades on pretty much any cloud platform) or even most simple things as liveness probes on Google Cloud.

Reply:
And we should forget about running in a sidecar.

Reply:
Can we put this on the agenda to discuss at standup tomorrow?  I don't think we actually ever talked this through.

Reply:
@bmanifold sgtm

Reply:
Went back and re-read the thread, gave it more thought, and I have a better idea of the concern @AndrewDryga has. Will be prepared to discuss at standup